patches/server/0717-Deobfuscate-stacktraces-in-log-messages-using-a-Rewr.patch

```diff
@@ -0,0 +1,363 @@
++                this.lines.add(line);
++            }
++
++            @Override
++            public void visitEnd() {
++                super.visitEnd();
++                lineMap.put(new Pair<>(this.name, this.descriptor), this.lines);
++            }
++        }
++        final ClassVisitor classVisitor = new ClassVisitor(Opcodes.ASM9) {
++            @Override
++            public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
++                return new LineCollectingMethodVisitor(name, descriptor);
++            }
++        };
++        try {
++            final ClassReader reader = new ClassReader(key.getName());
++            reader.accept(classVisitor, 0);
++        } catch (final IOException ex) {
++            throw new RuntimeException(ex);
++        }
++        return lineMap;
++    }
++
++    private @Nullable Pair<String, String> determineMethodForLine(final @NotNull Class<?> clazz, final int lineNumber) {
++        final Map<Pair<String, String>, IntList> lineMap = this.lineMapCache.computeIfAbsent(clazz, StacktraceDeobfuscatingRewritePolicy::buildLineMap);
++        for (final var entry : lineMap.entrySet()) {
++            final Pair<String, String> pair = entry.getKey();
++            final IntList lines = entry.getValue();
++            for (int i = 0, linesSize = lines.size(); i < linesSize; i++) {
++                final int num = lines.getInt(i);
++                if (num == lineNumber) {
++                    return pair;
++                }
++            }
++        }
++        return null;
++    }
++
++    private @NotNull String mappedFileName(final @NotNull String fullClassName) {
++        final int dot = fullClassName.lastIndexOf('.');
++        final String className = dot == -1
++            ? fullClassName
++            : fullClassName.substring(dot + 1);
++        final String rootClassName = className.split("\\$")[0];
++        return rootClassName + ".java";
++    }
++
++    @PluginFactory
++    public static @NotNull StacktraceDeobfuscatingRewritePolicy createPolicy() {
++        return new StacktraceDeobfuscatingRewritePolicy();
++    }
++
++    private record ClassMapping(
++        String obfName,
++        String mojangName,
++        Map<Pair<String, String>, MethodMapping> methodMappings
++    ) {
++    }
++
++    private record MethodMapping(
++        String obfName,
++        String mojangName,
++        String descriptor
++    ) {
++    }
++}
+diff --git a/src/main/resources/log4j2.xml b/src/main/resources/log4j2.xml
+index 67da1aa7a21622fb231d19dede3775a282a4a12e..f91d0df569b2f1d430ea5eee5f53779902a4f32c 100644
+--- a/src/main/resources/log4j2.xml
++++ b/src/main/resources/log4j2.xml
+@@ -29,15 +29,19 @@
+             </Policies>
+             <DefaultRolloverStrategy max="1000"/>
+         </RollingRandomAccessFile>
++        <Rewrite name="rewrite">
++            <StacktraceDeobfuscatingRewritePolicy />
```

Comment:
Can we make it possible to opt-out somehow?

Reply:
Can use a custom log4j config, but, could also add an option to make it do nothing if that's what you mean.

Reply:
Was thinking about that too. NMS devs will be super confused, so at least give them a easy opt out somehow, custom log4j config is a bit meh